Core/include/Acts/Utilities/TypeDispatcher.hpp

```diff
@@ -0,0 +1,193 @@
+/// it will look up and call the appropriate registered function.
+///
+/// @tparam base_t The base class type that will be the first parameter
+/// @tparam signature_t The function signature (e.g., return_t(Args...))
+template <typename base_t, typename signature_t>
+class TypeDispatcher;
+
+template <typename base_t, typename return_t, typename... args_t>
+class TypeDispatcher<base_t, return_t(args_t...)> {
+ public:
+  // Type aliases for frequently used template parameters
+  using base_type = base_t;
+  using return_type = return_t;
+  using self_type = TypeDispatcher<base_type, return_type(args_t...)>;
+
+  using function_signature = return_t(const base_t&, args_t...);
+  using function_type = std::function<function_signature>;
+
+  /// Default constructor
+  TypeDispatcher() = default;
+
+  /// Constructor that registers multiple function pointers with auto-detected
+  /// types
+  /// @param funcs Function pointers - derived types are auto-detected from their first parameters
+  template <typename... derived_types_t>
+  explicit TypeDispatcher(return_t (*... funcs)(const derived_types_t&,
+                                                args_t...))
+    requires(std::is_base_of_v<base_t, derived_types_t> && ...)
+  {
+    (registerFunction(funcs),
+     ...);  // Fold expression to register each function
+  }
+
+  /// Register a free function with explicit derived type
+  /// @tparam derived_t The derived type to associate the function with
+  /// @param func The function pointer
+  template <typename derived_t>
+    requires std::is_base_of_v<base_t, derived_t>
+  self_type& registerFunction(return_t (*func)(const derived_t&, args_t...)) {
+    std::type_index typeIdx(typeid(derived_t));
+
+    // Check if this exact type is already registered
+    if (m_functions.find(typeIdx) != m_functions.end()) {
+      throw std::runtime_error(
+          std::format("Function already registered for type: {}",
+                      boost::core::demangle(typeIdx.name())));
+    }
+
+    // Try to detect conflicts with existing registrations if the type is
+    // default constructible
+    if constexpr (std::is_default_constructible_v<derived_t>) {
+      derived_t tempObj{};
+      for (const auto& [existingTypeIdx, checker] : m_castCheckers) {
+        if (checker(tempObj)) {
+          throw std::runtime_error(
+              std::format("Registration conflict: type {} would be handled by "
+                          "existing function registered for type: {}",
+                          boost::core::demangle(typeIdx.name()),
+                          boost::core::demangle(existingTypeIdx.name())));
+        }
+      }
+    }
+
+    // Store a cast checker that tests if dynamic_cast<derived_t*> will work
+    m_castCheckers[typeIdx] = [](const base_t& obj) -> bool {
+      const auto* derived = dynamic_cast<const derived_t*>(&obj);
+      return derived != nullptr;
+    };
```

Comment:
Nice, and by this you don't have to unroll and cast, all that is required is that the registration and usage is synchronised.